tests/server_send.cpp

```diff
@@ -0,0 +1,349 @@
+			// following error:
+			// error: cannot bind packed field ... to int&
+			BH_LOG(log, DNET_LOG_DEBUG,
+					"ssend_test: "
+					"key: %s, backend: %d, user_flags: %llx, ts: %lld.%09lld, status: %d, size: %lld, "
+					"iterated_keys: %lld/%lld",
+				dnet_dump_id_len_raw(it->reply()->key.id, DNET_ID_SIZE, buffer),
+				(int)it->command()->backend_id,
+				(unsigned long long)it->reply()->user_flags,
+				(unsigned long long)it->reply()->timestamp.tsec, (unsigned long long)it->reply()->timestamp.tnsec,
+				(int)it->reply()->status, (unsigned long long)it->reply()->size,
+				(unsigned long long)it->reply()->iterated_keys, (unsigned long long)it->reply()->total_keys);
+#endif
+
+			copied++;
+		}
+
+		BH_LOG(log, DNET_LOG_NOTICE, "ssend_test: %s: dst_groups: %s, copied: %d",
+				id.to_string(),
+				print_groups(dst_groups), copied);
+
+		return copied;
+	};
+
+	int copied = 0;
+	std::vector<dnet_raw_id> ids = ssend_ids(s);
+	for (const auto &id: ids) {
+		copied += run_over_single_backend(s, id, dst_groups, iflags);
+	}
+
+	BOOST_REQUIRE_EQUAL(copied, num);
+}
+
+static void ssend_test_server_send(session &s, int num, const std::string &id_prefix, const std::string &data_prefix,
+		const std::vector<int> &dst_groups, uint64_t iflags)
+{
+	logger &log = s.get_logger();
+
+	s.set_trace_id(rand());
+	std::vector<std::string> keys;
+	for (int i = 0; i < num; ++i) {
+		std::string id = id_prefix + lexical_cast(i);
+		std::string data = data_prefix + lexical_cast(i);
+
+		ELLIPTICS_REQUIRE(res, s.write_data(id, data, 0));
+
+		keys.push_back(id);
+	}
+
+	BH_LOG(log, DNET_LOG_NOTICE, "%s: keys: %d, dst_groups: %s, starting copy",
+			__func__, num, print_groups(dst_groups));
+
+	//char buffer[2*DNET_ID_SIZE + 1] = {0};
+
+	int copied = 0;
+	auto iter = s.server_send(keys, iflags, dst_groups);
+	for (auto it = iter.begin(), iter_end = iter.end(); it != iter_end; ++it) {
+#if 0
+		// we have to explicitly convert all members from dnet_iterator_response
+		// since it is packed and there will be alignment issues and
+		// following error:
+		// error: cannot bind packed field ... to int&
+		BH_LOG(log, DNET_LOG_DEBUG,
+				"ssend_test: "
+				"key: %s, backend: %d, user_flags: %llx, ts: %lld.%09lld, status: %d, size: %lld, "
+				"iterated_keys: %lld/%lld",
+			dnet_dump_id_len_raw(it->reply()->key.id, DNET_ID_SIZE, buffer),
+			(int)it->command()->backend_id,
+			(unsigned long long)it->reply()->user_flags,
+			(unsigned long long)it->reply()->timestamp.tsec, (unsigned long long)it->reply()->timestamp.tnsec,
+			(int)it->reply()->status, (unsigned long long)it->reply()->size,
+			(unsigned long long)it->reply()->iterated_keys, (unsigned long long)it->reply()->total_keys);
+#endif
+
+		copied++;
+	}
+
+	BH_LOG(log, DNET_LOG_NOTICE, "%s: keys: %d, dst_groups: %s, copied total: %d",
+			__func__, num, print_groups(dst_groups), copied);
+
+	BOOST_REQUIRE_EQUAL(copied, num);
+}
+
+static bool ssend_register_tests(test_suite *suite, node &n)
```

Comment:
> What do you mean by potential bug? Is it data corruption, or maybe crash, which have been found here?

`Potential bug` - bug that was fixed within this review.

>This PR fixes couple of such things, and I didn't separate them because it is hard, and efforts on doing this do not worth amount of time needed. Apparently you do not count utility of the time discussing this things.

You do not count utility of our time when we try to keep this unrelated changes in mind and tries to analyze them separately.

> I'm tired wasting time with it. I've added this test. I do expect the same level of scrutiny for any other code now, I hope it will help to find bugs before next such hackathon.

Thanks. We always have the same level of scrutiny for all changes which could be merged into master.